# RFM Feature Engineering

This notebook computes Recency, Frequency, and Monetary (RFM) metrics from the cleaned retail dataset.
Feature engineering is performed in both SQL (canonical) and Pandas (validation), with quantile-scored RFM features saved for segmentation modeling.

---

RFM Metric Definitions:
- **Recency** – Days since the customer's last purchase (lower is better)
- **Frequency** – Total number of unique invoices per customer
- **Monetary** – Total revenue generated per customer (quantity × price)
- **R / F / M Scores** – Quartile-based scores (1–4) assigned via NTILE in SQL or `qcut`/`rank` in Pandas
- **RFM Score** – Composite score combining individual R, F, and M scores

In [1]:
import pandas as pd
import os
import getpass


from sqlalchemy import create_engine
from sqlalchemy.engine import URL
# Local src

from src.df_overview import df_overview
from src.sql_feature_engineering import create_database
from src.sql_feature_engineering import create_table
from src.sql_feature_engineering import load_data
from src.sql_feature_engineering import feature_engineering_sql
from src.pandas_feature_engineering import feature_engineering_pandas

print('Done')

Done


In [2]:
password = getpass.getpass('Enter database user password: ')

%load_ext sql

# connect to postgres and create the retail db
conn_str = f'postgresql+psycopg2://postgres:{password}@localhost:5432/postgres'
engine = create_engine(conn_str)
%sql engine

%sql --file ../sql/01_terminate_backend_sessions.sql
%sql --file ../sql/02_drop_database.sql
%sql --file ../sql/03_create_database.sql

# switch to retail db
conn_str = f'postgresql+psycopg2://postgres:{password}@localhost:5432/retail'
engine = create_engine(conn_str)
%sql engine

%sql --file ../sql/04_create_table.sql
# COPY needs psycopg2 directly
path_base_csv = '../data/02_processed/base_retail.csv'
load_data(path_base_csv, engine)

%sql --file ../sql/06_feature_engineering.sql

df_sql_featured = pd.read_sql(f"SELECT * FROM public.featured_retail", engine)
print(f"df_rfm loaded: {df_sql_featured.shape[0]} rows, {df_sql_featured.shape[1]} columns")
df_sql_featured.set_index('customer_id', inplace=True)
df_sql_featured.sort_index(inplace=True)
df_sql_featured

Running query in 'postgresql+psycopg2://postgres:***@localhost:5432/postgres'

Running query in 'postgresql+psycopg2://postgres:***@localhost:5432/postgres'

Running query in 'postgresql+psycopg2://postgres:***@localhost:5432/postgres'

Running query in 'postgresql+psycopg2://postgres:***@localhost:5432/retail'

Data loaded into 'base_retail'.


Running query in 'postgresql+psycopg2://postgres:***@localhost:5432/retail'

5847 rows affected.

df_rfm loaded: 5847 rows, 11 columns


,recency,frequency,monetary,recency_score,frequency_score,monetary_score,r_f_score,segment,return_ratio,avg_order_value
customer_id,,,,,,,,,,
12346,325,14,368.36,2,5,2,25,cant_lose,0.995250,26.311429
12347,1,8,4921.53,5,4,5,54,champions,0.000000,615.191250
12348,74,5,1658.40,3,3,4,33,need_attention,0.000000,331.680000
12349,18,4,3654.54,4,3,5,43,potential_loyalists,0.006565,913.635000
12350,309,1,294.40,2,1,2,21,hibernating,0.000000,294.400000
...,...,...,...,...,...,...,...,...,...,...
18283,3,22,2658.95,5,5,4,55,champions,0.000000,120.861364
18284,431,1,411.68,1,2,2,12,hibernating,0.000000,411.680000
18285,660,1,377.00,1,2,2,12,hibernating,0.000000,377.000000


In [ ]:
'''

# More as a pipeline script than a notebook, but if %sql magic is not available, this is how you can run the same steps in a Python script, just uncomment:

password = getpass.getpass('Enter database user password: ')

main_url = URL.create(
    drivername="postgresql+psycopg2",
    username="postgres",          # change if use different user
    password=password,
    host="localhost",             # change if connecting to different host
    port=5432,                    # change if connecting to different host
    database="postgres"          
)
# Create engine for the main database to create the new database
engine_main = create_engine(main_url)
path_base_csv = '../data/02_processed/base_retail.csv'

create_database(engine_main)

created_db_url = URL.create(
    drivername="postgresql+psycopg2",
    username="postgres",         
    password=password,
    host="localhost",
    port=5432,
    database='retail'          
)

# Create engine for the created database
engine_created_db = create_engine(created_db_url)

create_table(engine_created_db)

load_data(path_base_csv, engine_created_db)

feature_engineering_sql(engine_created_db)

df_sql_featured = pd.read_sql(f"SELECT * FROM public.featured_retail", engine_created_db)
print(f"df_rfm loaded: {df_sql_featured.shape[0]} rows, {df_sql_featured.shape[1]} columns")
df_sql_featured.set_index('customer_id', inplace=True)
df_sql_featured.sort_index(inplace=True)
df_sql_featured

'''

Executed: 01_terminate_backend_sessions.sql
Executed: 02_drop_database.sql
Executed: 03_create_database.sql
Database 'retail' created successfully.
Table 'base_retail' created in schema 'public'.
Data loaded into 'base_retail'.
Featured table 'public.featured_retail' created successfully.
df_rfm loaded: 5847 rows, 11 columns


,recency,frequency,monetary,recency_score,frequency_score,monetary_score,r_f_score,segment,return_ratio,avg_order_value
customer_id,,,,,,,,,,
12346,325,14,368.36,2,5,2,25,cant_lose,0.995250,26.311429
12347,1,8,4921.53,5,4,5,54,champions,0.000000,615.191250
12348,74,5,1658.40,3,3,4,33,need_attention,0.000000,331.680000
12349,18,4,3654.54,4,3,5,43,potential_loyalists,0.006565,913.635000
12350,309,1,294.40,2,1,2,21,hibernating,0.000000,294.400000
...,...,...,...,...,...,...,...,...,...,...
18283,3,22,2658.95,5,5,4,55,champions,0.000000,120.861364
18284,431,1,411.68,1,2,2,12,hibernating,0.000000,411.680000
18285,660,1,377.00,1,2,2,12,hibernating,0.000000,377.000000


In [4]:
'''
Pandas feature engineering CAN be used for testing purposes, but not implemented in this project.
The SQL version is the main implementation for feature engineering in this project.
if you have any issues with the SQL feature engineering, I built the pandas version for feature engineering, 
but comment out the SQL version and further testing code to avoid broken logic.
'''
df = pd.read_csv(path_base_csv)
df['invoice_date'] = pd.to_datetime(df['invoice_date'])
df_pd_featured = feature_engineering_pandas(df)
print(f"df_rfm loaded: {df_pd_featured.shape[0]} rows, {df_pd_featured.shape[1]} columns")
df_pd_featured.sort_index(inplace=True)
df_pd_featured

df_rfm loaded: 5847 rows, 10 columns


,recency,frequency,monetary,recency_score,frequency_score,monetary_score,r_f_score,segment,return_ratio,avg_order_value
customer_id,,,,,,,,,,
12346,325,14,368.36,2,5,2,25,cant_lose,NaN,26.311429
12347,1,8,4921.53,5,4,5,54,champions,NaN,615.191250
12348,74,5,1658.40,3,3,4,33,need_attention,NaN,331.680000
12349,18,4,3654.54,4,3,5,43,potential_loyalists,NaN,913.635000
12350,309,1,294.40,2,1,2,21,hibernating,NaN,294.400000
...,...,...,...,...,...,...,...,...,...,...
18283,3,22,2658.95,5,5,4,55,champions,NaN,120.861364
18284,431,1,411.68,1,2,2,12,hibernating,NaN,411.680000
18285,660,1,377.00,1,2,2,12,hibernating,NaN,377.000000


In [5]:
df_overview(df_sql_featured)

================================= Shape =================================
(5847, 10)
================================= Info =================================
<class 'pandas.core.frame.DataFrame'>
Index: 5847 entries, 12346 to 18287
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   recency          5847 non-null   int64  
 1   frequency        5847 non-null   int64  
 2   monetary         5847 non-null   float64
 3   recency_score    5847 non-null   int64  
 4   frequency_score  5847 non-null   int64  
 5   monetary_score   5847 non-null   int64  
 6   r_f_score        5847 non-null   object 
 7   segment          5847 non-null   object 
 8   return_ratio     5847 non-null   float64
 9   avg_order_value  5847 non-null   float64
dtypes: float64(3), int64(5), object(2)
memory usage: 502.5+ KB
None
================================= Columns =================================
Index(['recency', 'frequency', 'moneta

- 5847 customers for the analysis and modeling;
- no NaNs;
- correct dtypes;
- right features;
- no unexpected values for rfm calc;

In [6]:
df_overview(df_pd_featured)

================================= Shape =================================
(5847, 10)
================================= Info =================================
<class 'pandas.core.frame.DataFrame'>
Index: 5847 entries, 12346 to 18287
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   recency          5847 non-null   int64  
 1   frequency        5847 non-null   int64  
 2   monetary         5847 non-null   float64
 3   recency_score    5847 non-null   int64  
 4   frequency_score  5847 non-null   int64  
 5   monetary_score   5847 non-null   int64  
 6   r_f_score        5847 non-null   object 
 7   segment          5847 non-null   object 
 8   return_ratio     0 non-null      float64
 9   avg_order_value  5847 non-null   float64
dtypes: float64(3), int64(5), object(2)
memory usage: 502.5+ KB
None
================================= Columns =================================
Index(['recency', 'frequency', 'moneta

- some columns values are visually shown in exponent;
- r_f_score distribution is identical;
- segment destribution differs slightly;

In [7]:
# Testing dataframes on equality
pd.testing.assert_frame_equal(df_pd_featured, df_sql_featured, check_dtype=False)

AssertionError: DataFrame.iloc[:, 3] (column name="recency_score") are different

DataFrame.iloc[:, 3] (column name="recency_score") values are different (0.05131 %)
[index]: [12346, 12347, 12348, 12349, 12350, 12351, 12352, 12353, 12354, 12355, 12356, 12357, 12358, 12359, 12360, 12361, 12362, 12363, 12364, 12365, 12366, 12367, 12368, 12369, 12370, 12371, 12372, 12373, 12374, 12375, 12376, 12377, 12378, 12379, 12380, 12381, 12383, 12384, 12385, 12386, 12387, 12388, 12389, 12390, 12391, 12392, 12393, 12394, 12395, 12396, 12397, 12398, 12399, 12400, 12401, 12402, 12403, 12405, 12406, 12407, 12408, 12409, 12410, 12411, 12412, 12413, 12414, 12415, 12416, 12417, 12418, 12419, 12420, 12421, 12422, 12423, 12424, 12425, 12426, 12427, 12428, 12429, 12430, 12431, 12432, 12433, 12434, 12435, 12436, 12437, 12438, 12439, 12440, 12441, 12442, 12443, 12444, 12445, 12446, 12447, ...]
[left]:  [2, 5, 3, 4, 2, 2, 4, 2, 2, 2, 4, 4, 5, 5, 4, 2, 5, 3, 5, 2, 1, 5, 1, 1, 4, 3, 3, 2, 4, 5, 2, 2, 3, 3, 4, 5, 2, 4, 2, 2, 1, 5, 2, 3, 4, 1, 3, 3, 5, 1, 4, 4, 3, 1, 2, 2, 4, 3, 4, 4, 4, 3, 2, 1, 3, 3, 2, 4, 1, 5, 3, 2, 3, 5, 3, 5, 3, 3, 2, 5, 4, 5, 4, 4, 4, 5, 3, 3, 3, 5, 5, 1, 1, 2, 5, 1, 4, 4, 3, 2, ...]
[right]: [2, 5, 3, 4, 2, 2, 4, 2, 2, 2, 4, 4, 5, 5, 4, 2, 5, 3, 5, 2, 1, 5, 1, 1, 4, 3, 3, 2, 4, 5, 2, 2, 3, 3, 4, 5, 2, 4, 2, 2, 1, 5, 2, 3, 4, 1, 3, 3, 5, 1, 4, 4, 3, 1, 2, 2, 4, 3, 4, 4, 4, 3, 2, 1, 3, 3, 2, 4, 1, 5, 3, 2, 3, 5, 3, 5, 3, 3, 2, 5, 4, 5, 4, 4, 4, 5, 3, 3, 3, 5, 5, 1, 1, 2, 5, 1, 4, 4, 3, 2, ...]

- mismatch is small, less than 0.1% (0.05131%), the calculated data is almost identical;
- as I found, pandas qcut and rank have slighly different logic than NTILE in SQL, it probably causes this behavior;
- it is a better approach to do a full test to find out the exact causes, but the pandas version is here not primarily for testing purposes;

In [ ]:
pd_to_csv_path = os.path.join('..', 'data', '03_featured', 'featured_retail.csv')

os.makedirs(os.path.dirname(pd_to_csv_path), exist_ok=True)
df_sql_featured.to_csv(pd_to_csv_path)

print(f"Saved to: {pd_to_csv_path}")

Saved to: ..\data\03_featured\featured_retail.csv


- the SQL version is used as canonical and saved as featured csv;

## Summary of RFM Feature Engineering

### Approach:
- RFM metrics calculated using both **SQL (canonical)** and **Pandas (additional)**
- Return ratio and average order value are calculated for better segmentation and analysis to support RFM scores.

### Key Outputs:
- **5,847 customers** retained after cleaning, with no missing values and correct dtypes
- SQL and Pandas implementations produce near-identical results (**< 0.1% mismatch**, ~0.051%)
- Minor differences may be due to logic difference between NTILE and Pandas `qcut`

### Next Steps:
The featured dataset has been saved to `../data/03_featured/featured_rfm_retail.csv` and is ready for:
1. Post-feature EDA (RFM distributions, segment profiling, correlation analysis)
2. Customer segmentation modeling (clustering)